extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -0,0 +1,490 @@
+
+
+	public async refreshAccess(account: azdata.Account): Promise<azdata.Account> {
+		const response = await this.getCachedToken(account.key);
+		if (!response) {
+			account.isStale = true;
+			return account;
+		}
+
+		const refreshToken = response.refreshToken;
+		if (!refreshToken || !refreshToken.key) {
+			account.isStale = true;
+			return account;
+		}
+
+		try {
+			await this.refreshAccessToken(account.key, refreshToken);
+		} catch (ex) {
+			if (ex.message) {
+				await vscode.window.showErrorMessage(ex.message);
+			}
+			console.log(ex);
+		}
+		return account;
+	}
+
+
+	public async getSecurityToken(account: azdata.Account, azureResource: azdata.AzureResource): Promise<TokenResponse> {
+		const resource = this.resources.find(s => s.azureResourceId === azureResource);
+		if (!resource) {
+			return undefined;
+		}
+		const azureAccount = account as AzureAccount;
+		const response: TokenResponse = {};
+
+		for (const tenant of azureAccount.properties.tenants) {
+			let cachedTokens = await this.getCachedToken(account.key, resource.id, tenant.id);
+			// Check expiration
+			if (cachedTokens) {
+				const expiresOn = Number(this.memdb.get(`${account.key.accountId}_${tenant.id}_${resource.id}`));
+				const currentTime = new Date().getTime() / 1000;
+
+				if (!Number.isNaN(expiresOn)) {
+					const remainingTime = expiresOn - currentTime;
+					// expired
+					if (remainingTime < 5 * 60) {
+						cachedTokens = undefined;
+					}
+				} else {
+					// No expiration date, assume expired.
+					cachedTokens = undefined;
+					console.log('Assuming expired token');
+				}
+
+			}
+
+			// Refresh
+			if (!cachedTokens) {
+
+				const baseToken = await this.getCachedToken(account.key);
+				if (!baseToken) {
+					return undefined;
+				}
+
+				await this.refreshAccessToken(account.key, baseToken.refreshToken, tenant, resource);
+				cachedTokens = await this.getCachedToken(account.key, resource.id, tenant.id);
+				if (!cachedTokens) {
+					return undefined;
+				}
+			}
+			const { accessToken } = cachedTokens;
+			response[tenant.id] = {
+				token: accessToken.token,
+				key: accessToken.key,
+				tokenType: 'Bearer'
+			};
+		}
+
+		if (azureAccount.properties.subscriptions) {
+			azureAccount.properties.subscriptions.forEach(subscription => {
+				response[subscription.id] = {
+					...response[subscription.tenantId]
+				};
+			});
+		}
+
+		return response;
+	}
+
+	public async clearCredentials(account: azdata.AccountKey): Promise<void> {
+		try {
+			return this.deleteAccountCache(account);
+		} catch (ex) {
+			const msg = localize('azure.cacheErrrorRemove', "Error when removing your account from the cache.");
+			console.error('Error when removing tokens.', ex);
+			throw new Error(msg);
+		}
+	}
+
+	protected toBase64UrlEncoding(base64string: string) {
+		return base64string.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_'); // Need to use base64url encoding
```

Comment:
I would suggest `Buffer.from(base64string).toString('base64')`

Reply:
Hmm I think I had a reason that I couldn't use that but it's been so long now. I'll test it again. Thanks!

Reply:
Ah yeah, 

![image](https://user-images.githubusercontent.com/1465785/77199172-a8102580-6aa5-11ea-8e5a-5cbc3ddcbe9f.png)

There doesn't seem to be a built in method to URL Encode base64 strings.

Reply:
Base64 includes `/` and `+` and `=`. The `=` isn't part of base64 but its used for padding. Essentially I need to get rid of these and replace them with `-`, `_` respectively and remove the padding.

Reply:
Makes sense, came across this too - https://github.com/nodejs/node/issues/26512